development-docs/DEBUGGING.md

```diff
@@ -0,0 +1,188 @@
+
+We can enable remote debugging by adding the `JAVA_OPTS` environment variable to `strimzi-cluster-operator` container in this file:
+
+          env:
+            ...
+            - name: JAVA_OPTS
+              value: "-agentlib:jdwp=transport=dt_socket,server=y,suspend=y,address=5005"
+
+
+Let's also increase the livenessProbe timeout:
+
+          livenessProbe:
+            initialDelaySeconds: 3600
+            timeoutSeconds: 3600
+
+Apply the modified file:
+
+    kubectl apply -f install/cluster-operator/060-Deployment-strimzi-cluster-operator.yaml
+
+We also need to open the port for localhost access:
+
+    kubectl port-forward $(oc get pod | grep strimzi-cluster-operator | awk '{printf $1}') 5006:5005
+
+Start tailing the operator pod:
+
+    oc logs $(oc get pod | grep strimzi-cluster-operator | awk '{printf $1}') -f 
+
+You can now start the remote debug session from your IDE.
+
+
+## Remote debugging the Kafka Broker
+
+Kafka Broker pods are defined through 'my-cluster-kafka' StatefulSet which is created by Strimzi Kafka Operator.
+
+In order to activate the remote debugging we need to get the Kafka Broker to run with remote debugging agent activated.
+
+This can be achieved by adding the following to your Kafka CR cluster definition:
+
+```
+
+spec:
+  kafka:
+    ...
+
+    livenessProbe:
+      initialDelaySeconds: 3600
+      timeoutSeconds: 3600
+    template:
+      kafkaContainer:
+        env:
+        - name: KAFKA_DEBUG
+          value: "y"
+        - name: DEBUG_SUSPEND_FLAG
+          value: "y"
+        - name: JAVA_DEBUG_PORT
+          value: "5005"
+
+```
+
+We also need to open the specified port (5005 in this case) to be reachable from your host machine:
+
+    kubectl port-forward my-cluster-kafka-0 5005
+
+If you want to map to a different local port use the LOCAL:REMOTE format, for example:
+
+    kubectl port-forward my-cluster-kafka-0 8787:5005
+
+All you now have to do is deploy the new Kafka cluster definition, configure remote debugging in your favourite IDE, telling it to connect to localhost:5005, and start your debug session.
+
+With `DEBUG_SUSPEND_FLAG` set to 'y', the Kafka Broker process will wait during startup for remote degugger (IDE) to connect before continuing with JVM start up.
+
+In Kafka CR we have adjusted the probes to prevent Kubernetes from killing the broker whose execution is suspended due to a breakpoint.
+
+Before remotely connecting with IDE you'll want to start tailing the `my-cluster-kafka-0` broker pod:
+
+    kubectl logs my-cluster-kafka-0 -f
+
+You should see the following as the last line:
+
+    Listening for transport dt_socket at address: 5005
+
+You can now start a remote debug session from your IDE.
+
+Note: Apache Kafka is written in a way that it detects stalled request processing threads and may exit the JVM that appears to have stalled.
```

Comment:
Is there any option I can set to make this wait longer?

Reply:
Maybe using 'Stop all threads' breakpoint option would help - if all thread are suspended the one doing the shutdown can't trigger it. But it would trigger it as soon as you continue let the thread continue normal execution (the green play button). Maybe there are some low level settings to prolong the tolerated request thread stall. Maybe check the upstream Kafka docs